contracts/external/callees/Trade0x.sol

```diff
@@ -0,0 +1,161 @@
+/**
+ * SPDX-License-Identifier: UNLICENSED
+ */
+
+pragma solidity 0.6.10;
+
+pragma experimental ABIEncoderV2;
+
+import {CalleeInterface} from "../../interfaces/CalleeInterface.sol";
+import {IZeroXExchange} from "../../interfaces/ZeroXExchangeInterface.sol";
+import {SafeERC20} from "../../packages/oz/SafeERC20.sol";
+import {ERC20Interface} from "../../interfaces/ERC20Interface.sol";
+import {WETH9} from "../canonical-weth/WETH9.sol";
+
+/**
+ * @author Opyn Team
+ * @title Trade0x
+ * @notice callee contract to trade on 0x.
+ */
+
+contract Trade0x is CalleeInterface {
+    using SafeERC20 for ERC20Interface;
+    IZeroXExchange public exchange;
+    ERC20Interface public weth;
+    address public controller;
+    address public assetProxy;
+    address public staking;
+    ///@dev 0x portocal fee to fill 1 order
+    uint256 private PORTOCAL_FEE_BASE = 70000;
+
+    constructor(
+        address _exchange,
+        address _assetProxy,
+        address _weth,
+        address _staking,
+        address _controller
+    ) public {
+        exchange = IZeroXExchange(_exchange);
+        assetProxy = _assetProxy;
+        weth = ERC20Interface(_weth);
+        weth.safeApprove(_assetProxy, uint256(-1));
+        weth.safeApprove(_staking, uint256(-1));
+        controller = _controller;
+    }
+
+    event Trade(
+        address indexed taker,
+        address indexed maker,
+        address takerAsset,
+        address makerAsset,
+        uint256 takerAmount,
+        uint256 makerAmount
+    );
+
+    /**
+     * @dev fill 0x order
+     * @param _sender the original sender who wants to trade on 0x
+     * @param _data abi-encoded order, fillamount, signature and feePayer. fee payer is the address we pull weth from.
+     */
+
+    function callFunction(address payable _sender, bytes memory _data) external override {
+        require(msg.sender == controller, "sender not controller");
+        _directlyTrade(_sender, _data);
+    }
+
+    function _directlyTrade(address payable _sender, bytes memory _data) internal {
+        (
+            IZeroXExchange.Order[] memory order,
+            uint256[] memory takerAssetFillAmount,
+            bytes[] memory signature,
+            address feePayer
```

Comment:
Just to make clear here, we can't let the `feePayer` max approve this Contract as any other user can use the same `feePayer` address and pay fee from that wallet, if it have enough WETH available, right ?

Reply:
Yeah, so if the fee payer is us, then its fine. but yes, otherwise this is an issue and every time the user makes a transaction they have to approve this contract to spend weth.

Reply:
why can't we check relationship between `_feePayer` and `_sender` in this function? In most cases, this `_feePayer` will be `_sender`:
1. User call Controller directly, `_sender` = user address, and `_feePayer` = user address
2. User call PayableProxy => Controller, `_sender` = payable proxy, `feePayer` should also be the payable proxy. (User send eth to payableProxy => convert ETH to WETH and stay in the payable proxy contract => Being pulled by 0xCallee).

And if _sender != _feePayer, `require Controller.isOperator(_feePayer, _sender) === true`, in case user want to use another operator but still let our callee just pull weth from their wallet.